Enhanced Heartbeat Telemetry

andrewvc opened the pull request:
Add heartbeat telemetry. We now track:

 1. The total number of active monitors
 2. The total number of active monitors per protocol
 3. The total number of checked endpoints per protocol

One note for reviewers, I changed the signature of the `create` function used by heartbeat plugins to return the number of `endpoint`s for the given monitor.

An alternate approach I considered was refactoring from the FP style for plugins to a more OO style where the stats could be pulled *after* creation by navigating a map of monitor objects. I considered this to be a more minimal change, but would welcome some feedback.

andrewvc commented on heartbeat/monitors/monitor.go:
```diff
@@ -30,8 +30,38 @@ import (
 )
 
+var teleRegistry *monitoring.Registry = monitoring.Default.NewRegistry("heartbeat")
+var httpRegistry *monitoring.Registry = teleRegistry.NewRegistry("heartbeat.http")
```
@urso I initially thought I could nest by writing:

```go
var teleRegistry *monitoring.Registry = monitoring.Default.NewRegistry("heartbeat")
var httpRegistry *monitoring.Registry = teleRegistry.NewRegistry("http")
```

but discovered that despite calling `NewRegistry` on an existing namespaced registry I still needed the full path. Is this by design? If so, what is the intended purpose of being able to chain `NewRegistry`? Should I refactor this to always invoke `NewRegistry` on `monitoring.Default`?

urso replied:
Why do you need the full path? Where did you see things going wrong?

When calling `(*Regsitry).NewRegistry(name)`, then it adds the new registry as a child to the parent with `name` as field. This way you build a tree.
The collectors/representation is not aware of dots like `heartbeat.http`. Actually you should not use dots for names in your registry. Depending on the metrics collector in use you might get events like:

```
{
  "heartbeat": {
    "heartbeat.http": {
      ...
    }
  },
  ...
}
```

andrewvc commented on heartbeat/monitors/active/tcp/tcp.go:
```diff
@@ -47,33 +47,32 @@ type connURL struct {
 	}
 
-	endpoints, err := collectHosts(&config, defaultScheme)
+	schemeHosts, err := collectHosts(&config, defaultScheme)
```
This was unclear before. This is really a map of `scheme(str) -> []host(str)`

ruflin commented on heartbeat/monitors/active/http/http.go:
```diff
@@ -58,44 +58,44 @@ func create(
 			jobs[i], err = newHTTPMonitorIPsJob(&config, url, tls, enc, body, validator)
 			if err != nil {
-				return nil, err
+				return nil, -1, err
```
Why did you pick -1 and not 0?

andrewvc replied:
I'm not partial to -1 or 0. I picked -1  because it's obviously a nonsense value in the error case. In any event, in all these situations the value should be discarded by any callers.

ruflin replied:
My thinking: The method returns the number of endpoints it created. In the case of error this is 0 and not -1. Lets assume I always count the value even in the error case, the results would still be correct.

ruflin commented on heartbeat/monitors/active/icmp/icmp.go:
```diff
@@ -37,10 +37,10 @@ var debugf = logp.MakeDebug("icmp")
 	config := DefaultConfig
 	if err := cfg.Unpack(&config); err != nil {
-		return nil, err
+		return nil, -1, err
```
Same question as above, why -1 instead of 0?

ruflin commented on heartbeat/monitors/mocks_test.go:
```diff
@@ -99,10 +99,10 @@ func createMockJob(name string, cfg *common.Config) ([]Job, error) {
-	return pluginBuilder{"test", ActiveMonitor, func(s string, config *common.Config) ([]Job, error) {
+	return pluginBuilder{"test", ActiveMonitor, func(s string, config *common.Config) ([]Job, int, error) {
 		c := common.Config{}
 		j, err := createMockJob("test", &c)
```
I think this breaks the tests because there is not monitor type `test`.

andrewvc replied:
This is now fixed since we only log instead of panic in this case.

andrewvc replied:
This is now further improved by our dynamic creation of metrics, so this scenario is no longer possible.

ruflin commented on heartbeat/monitors/monitor.go:
```diff
@@ -227,6 +261,15 @@ func (m *Monitor) Start() {
+	teleStats.monitors.Inc()
+
+	if stats, ok := teleStats.protocols[m.name]; !ok {
+		panic(fmt.Sprintf("Unknown protocol for monitor stats: %s", m.name))
```
Should we really panic if can't to the stats? Assuming someone adds a new protocol, forgets to update the list and we do a release I would expect that we log an error but not stop the application because of it.

On the other hand having a panic makes it easy to spot the problem early on.

andrewvc replied:
I tend to favor panic for "This should never happen" sorts of things.

The benefit is that if we add a new protocol we won't forget stats support. In this case, I think its OK, because I can't see how we could go through a dev/test/release cycle without seeing the panic.

andrewvc replied:
OTOH, this breaks with the mock "test" monitor. Maybe it should jus be a log message.

WDYT? The other approach would be to add "test " as a protocol (or going further, to inject the map of monitors, but that seems like overkill).

ruflin commented on heartbeat/monitors/plugin_test.go:
```diff
@@ -194,43 +192,6 @@ func Test_pluginsReg_String(t *testing.T) {
 	}
 }
 
-func Test_pluginBuilder_create(t *testing.T) {
```
What happened to these?

andrewvc replied:
This whole file should be deleted. It was auto-generated, but never filled in. This particular test no longer compiled, so I just deleted it. I'll delete the whole file in a subsequent PR.

ruflin commented on heartbeat/tests/system/heartbeat.py:
```diff
@@ -32,3 +33,43 @@ def do_GET(self):
         thread.start()
 
         return server
+
```
+1 on having this in the heartbeat.py file

ruflin commented:
Seems like Jenkins is now failing because "address already in use", meaning some services are not stopped after testing?

andrewvc commented:
@ruflin I think I may have fixed that in 0e20f0a , waiting for a test run. I remembered to tear down the server, but I forgot to kill heartbeat itself.

urso commented on heartbeat/monitors/monitor.go:
```diff
@@ -30,8 +30,38 @@ import (
+		"tcp":  newProtocolStats(teleRegistry.NewRegistry("tcp")),
+		"icmp": newProtocolStats(teleRegistry.NewRegistry("icmp")),
+	},
+}
```
I would prefer for new code not to rely on globals.

andrewvc commented on heartbeat/monitors/plugin_test.go:
```diff
@@ -1,210 +0,0 @@
-// Licensed to Elasticsearch B.V. under one or more contributor
```
This file was accidentally committed in a past PR and was giving compiler trouble here.

andrewvc commented on heartbeat/monitors/plugin.go:
```diff
@@ -129,7 +149,7 @@ func (r *pluginsReg) monitorNames() []string {
 }
 
-func (e *pluginBuilder) create(cfg *common.Config) ([]Job, error) {
+func (e *pluginBuilder) create(cfg *common.Config) (jobs []Job, endpoints int, err error) {
```
explicit names because `endpoints int` is more explanatory than `int`

ruflin commented:
For the reporting. I don't think we should report the exact same data to stats and state. Also the data on the state side will need a heartbeat prefix as this data is very specific to heartbeat compared to the other data we have in state.

I'm thinking on the stats side we could report the total number of starts and stops of monitor and the one we use for the state is the gauge (start - stop).

andrewvc commented:
@ruflin can you outline your thinking here as far as why we'd want counters instead of a gauge? I'm having trouble seeing the use cases where that'd be easier to use. That said, I'm not familiar how people use our stats API in the real world, would love to hear your opinion.

Also, thanks for catching the missing prefix, I just added it.

ruflin commented:
The counter with a derivative has the advantage that is accurate independent of the time frame selected and also if some events were not reported. 

Most important for me in this PR is that we report the state which I agree should be the gauge. Lets simplify the PR to only to state and add stats later?

dov0211 commented:
@ruflin @andrewvc do we have the time to merge it today?

dov0211 commented:
@ruflin @andrewvc do we have the time to merge it today?

andrewvc commented:
@ruflin You're right. I actually used counters for this reason in Logstash's monitoring solution and forgot about it.

As you point out, we can totally get both of these numbers with counters. To do so we need to also send an UUID associated with a given process invocation (sort of a global PID) to bucket the counters by. Do we currently have this in the telemetry data?

If you know, that'd be great, in the meantime I'll dig in and see what we have as far as UUIDs there.

andrewvc commented:
@ruflin ok, I switched to counters in 998d27a, and removed metrics, however, now I need find a way to test telemetry since AFAIK there's externally observable behavior. I can write some unit tests of course.

This is probably a good thing, since we should test both.

andrewvc commented:
@ruflin per our discussion:

1. The state data now uses gauges
2. The stats data now uses counters

Test suite has been updated as well. Tests pass locally, waiting for CI to confirm.

ruflin commented:
Seems like CI is breaking because of reuse of the `monitors` key?

andrewvc commented:
@ruflin fixed! When I refactored I accidentally was re-initializing the global metrics once-per plugin.

andrewvc commented:
The one failure is unrelated:

```
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/private/var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-darwin/beat/libbeat/label/macosx/src/github.com/elastic/beats/libbeat/tests/system/test_keystore.py", line 39, in test_keystore_with_present_key
    proc.check_kill_and_wait()
  File "/private/var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-darwin/beat/libbeat/label/macosx/src/github.com/elastic/beats/libbeat/tests/system/beat/beat.py", line 97, in check_kill_and_wait
    return self.check_wait(exit_code=exit_code)
  File "/private/var/lib/jenkins/workspace/elastic+beats+pull-request+multijob-darwin/beat/libbeat/label/macosx/src/github.com/elastic/beats/libbeat/tests/system/beat/beat.py", line 86, in check_wait
    exit_code, actual_exit_code)
AssertionError: Expected exit code to be 0, but it was -15
```
